List and keys translation (#58)
reactjs/pl.reactjs.org: content/docs/lists-and-keys.md

---
id: lists-and-keys
title: Listy i klucze
permalink: docs/lists-and-keys.html
prev: conditional-rendering.html
next: forms.html
---
 
Na początku, przyjrzymy się jak przekształca się listy w JavaScript.
 

> W kodzie poniżej, użyliśmy funkcji [`map()`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/map) aby wziąć tablicę `liczb`i podwoić och wartość. Następnie przypisujemy nową tablice zwróconą z funkcji `map()` do zmiennej `doubled` i wyświetlamy ją:

Reviewer: Żeby zachować spójność czasów, proponuję: "użyliśmy funkcji [...]. Następnie przypisaliśmy nową tablicę...".

> W kodzie poniżej, użyliśmy funkcji [`map()`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/map) aby wziąć tablicę `liczb`i podwoić och wartość. Następnie przypisujemy nową tablice zwróconą z funkcji `map()` do zmiennej `doubled` i wyświetlamy ją:

Author: Poprawki naniesione.

> W kodzie poniżej, użyliśmy funkcji [`map()`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/map) aby wziąć tablicę `liczb`i podwoić och wartość. Następnie przypisujemy nową tablice zwróconą z funkcji `map()` do zmiennej `doubled` i wyświetlamy ją:

Author: Naniosłem obie sugestie. 👍 Thanks

```javascript{2}
const numbers = [1, 2, 3, 4, 5];
const doubled = numbers.map((number) => number * 2);
console.log(doubled);
```
 

> Ten kod wyświetla w konsoli `[2, 4, 6, 8, 10]`.

Reviewer: ```suggestion
Kod ten wyświetla w konsoli `[2, 4, 6, 8, 10]`.
```

> W Reakcie, przekształcanie tablic do list [elementów](/docs/rendering-elements.html) jest niemalże identyczne

Reviewer: ```suggestion
W Reakcie przekształcanie tablic w listy [elementów](/docs/rendering-elements.html) przebiega niemalże identycznie.
```

 
### Wyświetlanie wielu komponentów {#rendering-multiple-components}
 

> Możesz zbudować kolekcje elementów i [dodać je do JSX](/docs/introducing-jsx.html#embedding-expressions-in-jsx) używając  klamrowych nawiasów `{}`.

Reviewer: ```suggestion
Możesz zbudować kolekcje elementów i [dodać je do JSX](/docs/introducing-jsx.html#embedding-expressions-in-jsx), używając nawiasów klamrowych `{}`.
```

> Poniżej, iterujemy tablicę `liczb` używając funkcji [`map()`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/map) JavaScript. Zwracamy element `<li>` dla każdego elementu tablicy. Na koniec, przypisujemy wynikową tablie do `listItems`:

Reviewer 2: ```suggestion
Poniżej iterujemy tablicę liczb `numbers` używając javascriptowej funkcji [`map()`](https://developer.mozilla.org/pl/docs/Web/JavaScript/Referencje/Obiekty/Array/map). Zwracamy element `<li>` dla każdego elementu tablicy. Na koniec przypisujemy wynikową tablicę do `listItems`:
```

> Poniżej, iterujemy tablicę `liczb` używając funkcji [`map()`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/map) JavaScript. Zwracamy element `<li>` dla każdego elementu tablicy. Na koniec, przypisujemy wynikową tablie do `listItems`:

Reviewer: ```suggestion
Poniżej iterujemy tablicę liczb `numbers` używając javascriptowej funkcji [`map()`](https://developer.mozilla.org/pl/docs/Web/JavaScript/Referencje/Obiekty/Array/map). Zwracamy element `<li>` dla każdego elementu tablicy. Na koniec przypisujemy powstałą w ten sposób tablicę do zmiennej `listItems`:
```

);
```
 

> [**Wypróbuj na CodePen**](https://codepen.io/gaearon/pen/GjPyQr?editors=0011)

Reviewer: ```suggestion
[**Przetestuj kod na CodePen**](https://codepen.io/gaearon/pen/GjPyQr?editors=0011)
```
Zgodnie z naszym Glosariuszem na Slacku.

> Ten kod wyświetla wypunktowaną listę numerów pomiędzy 1 i 5.

Reviewer 2: ```suggestion
Ten kod wyświetla wypunktowaną listę liczb od 1 do 5.
```

 
### Podstawowy komponent listy {#basic-list-component}
 

> Zazwyczaj będziesz chciał wyświetlić listę wewnątrz [komponentu](/docs/components-and-props.html).

Reviewer 2: ```suggestion
Zazwyczaj będziesz wyświetlać listy wewnątrz [komponentu](/docs/components-and-props.html).
```

> Możemy zmienić poprzedni przykład w komponent który akceptuje tablicę `liczb` i zwraca listę elementów.

Reviewer 2: ```suggestion
Możemy zmienić poprzedni przykład w komponent który akceptuje tablicę liczb `numbers` i zwraca listę elementów.
```

> Możemy zmienić poprzedni przykład w komponent który akceptuje tablicę `liczb` i zwraca listę elementów.

Reviewer: ```suggestion
Możemy przekształcić poprzedni przykład w komponent, który akceptuje tablicę liczb `numbers` i zwraca listę elementów.
```

);
```
 

> Kiedy uruchomisz ten kod, dostaniesz ostrzeżenie o tym że właściwość klucza powinna zostać dostarczona do elementów listy. "Klucz" jest specjalnym atrybutem o typie łańcucha znaków, który musisz dodać podczas tworzenia elementów listy. O tym dlaczego jest to ważne opowiemy w następnej sekcji.

Reviewer 2: ```suggestion
Kiedy uruchomisz ten kod, dostaniesz ostrzeżenie o tym, że właściwość klucza powinna zostać dostarczona do elementów listy. „Klucz” (ang. *key*) jest specjalnym atrybutem o typie łańcucha znaków, który musisz dodać podczas tworzenia elementów listy. O tym dlaczego jest to ważne opowiemy w następnej sekcji.
```

Reviewer: ```suggestion
Kiedy uruchomisz powyższy kod, dostaniesz ostrzeżenie o tym, że do elementów listy należy dostarczyć właściwość klucza. „Klucz” (ang. *key*) jest specjalnym atrybutem o typie łańcucha znaków, który musisz dodać podczas tworzenia elementów listy. O tym, dlaczego jest to ważne, opowiemy w następnej sekcji.
```

Author: @jfrej @jakubdrozdek Naniosłem zmiany łącząc obie sugestie. 👍 Dzięki

> Przypiszmy `klucz` do naszych elementów listy wewnątrz `numbers.map()` czym naprawimy problem z brakującym kluczem.

Reviewer 2: ```suggestion
Przypiszmy właściwość `key` do naszych elementów listy wewnątrz `numbers.map()`, czym naprawimy problem brakującego klucza.
```

);
```
 

> [**Wypróbuj na CodePen**](https://codepen.io/gaearon/pen/jrXYRR?editors=0011)

Reviewer: ```suggestion
[**Przetestuj kod na CodePen**](https://codepen.io/gaearon/pen/jrXYRR?editors=0011)
```

 
## Klucze {#keys}
 

> Klucze pomagają Reaktowi zidentyfikować które elementy uległy zmianie, zostały dodane lub usunięte. Klucze powinny zostać nadane elementom wewnątrz tablic aby zyskały stabilną tożsamość:

Reviewer 2: ```suggestion
Klucze pomagają Reaktowi zidentyfikować, które elementy uległy zmianie, zostały dodane lub usunięte. Klucze powinny zostać nadane elementom wewnątrz tablicy, aby elementy zyskały stabilną tożsamość:
```

);
```
 

> Najlepszym sposobem wyboru klucza jest użycie unikatowego ciągu znaków który jednoznacznie identyfikuje element listy spośród jego rodzeństwa. Jako klucze często będziesz używał identyfikatór (*ang.* IDs) ze swoich danych:

Reviewer 2: ```suggestion
Najlepszym sposobem wyboru klucza jest użycie unikatowego ciągu znaków, który jednoznacznie identyfikuje element listy spośród jego rodzeństwa. Jako kluczy często będziesz używać identyfikatorów (ang. *IDs*) ze swoich danych:
```

);
```
 

> Kiedy nie będziesz miał stabilnych identyfikatorów dla wyświetlanych elementów, w skrajnym przypadku możesz użyć indeksu jako klucz:

Reviewer 2: ```suggestion
Kiedy nie będziesz miał stabilnych identyfikatorów dla wyświetlanych elementów, w skrajnym przypadku możesz użyć indeksu jako klucza:
```

> Kiedy nie będziesz miał stabilnych identyfikatorów dla wyświetlanych elementów, w skrajnym przypadku możesz użyć indeksu jako klucz:

Reviewer: ```suggestion
Jeśli w danych nie masz stabilnych identyfikatorów dla wyświetlanych elementów, w skrajnym przypadku do określenia klucza możesz użyć indeksu elementu tablicy:
```
Starajmy się pisać bezosobowo.

 
```js{2,3}
const todoItems = todos.map((todo, index) =>

>   // Rób to tylko w przypadku gdy elementy nie mają stabilnego identyfikatorów

Reviewer 2: ```suggestion
  // Rób to tylko w przypadku, gdy elementy nie mają stabilnych identyfikatorów
```

  <li key={index}>
    {todo.text}
  </li>
);
```
 

> Nie zalecamy używania indeksów jako klucze jeżeli kolejność elementów może ulec zmianie. Może to negatywnie wpłynąć na wydajność i spowodować problemy ze stanem komponentu. Sprawdź artykuł Robin Pokorny na temat [szczegółowego wyjaśnienia negatywnych aspektów używania indeksu jako klucz](https://medium.com/@robinpokorny/index-as-a-key-is-an-anti-pattern-e0349aece318). Jeżeli nie zdecydujesz się jawnie ustawić kluczy dla elementów listy, React domyślnie użyje indeksów jako klucze.

Reviewer 2: ```suggestion
Nie zalecamy używania indeksów jako kluczy, jeżeli kolejność elementów może ulec zmianie. Może to negatywnie wpłynąć na wydajność i spowodować problemy ze stanem komponentu. Sprawdź artykuł Robina Pokorny'ego ze [szczegółowym wyjaśnieniem negatywnych aspektów używania indeksu jako klucza](https://medium.com/@robinpokorny/index-as-a-key-is-an-anti-pattern-e0349aece318). Jeżeli nie zdecydujesz się jawnie ustawić kluczy dla elementów listy, React domyślnie użyje indeksów jako kluczy.
```

> Jeżeli chciałbyś dowiedzieć się wiecej, to tutaj znajdziesz [szczegółowe wyjaśnienie dlaczego klucze są niezbędne](/docs/reconciliation.html#recursing-on-children).

Reviewer 2: ```suggestion
Jeżeli chcesz dowiedzieć się więcej, tutaj znajdziesz [szczegółowe wyjaśnienie dlaczego klucze są niezbędne](/docs/reconciliation.html#recursing-on-children).
```

> Jeżeli chciałbyś dowiedzieć się wiecej, to tutaj znajdziesz [szczegółowe wyjaśnienie dlaczego klucze są niezbędne](/docs/reconciliation.html#recursing-on-children).

Reviewer: ```suggestion
Jeżeli chcesz dowiedzieć się więcej, tutaj znajdziesz [szczegółowe wyjaśnienie, dlaczego klucze są niezbędne](/docs/reconciliation.html#recursing-on-children).
```

> ### Wydrębnianie komponentów z kluczami {#extracting-components-with-keys}

Reviewer 2: ```suggestion
### Wyodrębnianie komponentów z kluczami {#extracting-components-with-keys}
```
lub:
```suggestion
### Wyciąganie komponentów z kluczami {#extracting-components-with-keys}
```
bo tak zostało to przetłumaczone na stronie o komponentach: https://github.com/reactjs/pl.reactjs.org/pull/45/files#diff-becb6cc71f7755db88023cfd07ca9a1fR125

Osobiście wolę "wyodrębnianie" i zmieniłbym w obu miejscach.

Author: @jakubdrozdek Nie orientuje się jak wyglądała tam sprawa z `Wyciąganie` vs `Wyodrębnianie` ?
Osobiście jestem za @jfrej i wybrałbym `Wyodrębnianie` oraz naniósł poprawkę pod tym lub osobnym PR do https://github.com/reactjs/pl.reactjs.org/pull/45/files#diff-becb6cc71f7755db88023cfd07ca9a1fR125.

Daj znać co o tym sądzisz.

Reviewer: Masz rację. Ja się zajmę poprawką w tamtym pliku. Tutaj niech zostaną rzeczy związane z Listami i kluczami.

Reviewer: Poprawki wrzucone do osobnego PR-a: #70

 
Klucze mają sens tylko w kontekście otaczającej tablicy.
 

> Dla przykładu, jeżeli [wyodrębnisz](/docs/components-and-props.html#extracting-components) komponent `ListItem`, powinieneś trzymać klucz na elementach `<ListItem />`wewnątrz tablicy zamiast na elemencie `<li>` wewnątrz elementu `ListItem`.

Reviewer 2: ```suggestion
Dla przykładu, jeżeli [wyodrębnisz](/docs/components-and-props.html#extracting-components) komponent `ListItem`, powinieneś trzymać klucz na elementach `<ListItem />` wewnątrz tablicy zamiast na elemencie `<li>` wewnątrz komponentu `ListItem`.
```

Reviewer: ```suggestion
Dla przykładu, jeżeli [wyodrębnisz](/docs/components-and-props.html#extracting-components) komponent `ListItem`, trzymaj klucz na elementach `<ListItem />` wewnątrz tablicy zamiast na elemencie `<li>` wewnątrz elementu `ListItem`.
```

Author: Done.

> **Przykłaď: Niepoprawne użycie klucza**

Reviewer 2: ```suggestion
**Przykład: niepoprawne użycie klucza**
```

Reviewer: Według mnie to jest tytuł pewnej sekcji, więc powinniśmy zacząć go wielką literą.

Author: Też bym zostawił z dużej litery, chociaż nie jestem pewien czy tak jest poprawnie.
Może @marcinostrouch albo @cvqprs będzie wiedział? 🎉

Reviewer 3: +1 dla wielkiej litery

);
```
 

> **Przykład: Poprawne użycie klucza**

Reviewer 2: ```suggestion
**Przykład: poprawne użycie klucza**
```

Reviewer: Jestem za napisaniem "Poprawne".

Reviewer 3: +1 podobnie jak w poprzednim przykladzie

 
```javascript{2,3,9,10}
function ListItem(props) {

>   // Dobrze! Nie ma potrzeby ustawiania klucza tutaj:

Reviewer 2: Dla konsekwencji można przetłumaczyć jak wcześniej:
```suggestion
  // Dobrze! Nie ma potrzeby definiowania klucza tutaj:
```

);
```
 

> [**Wypróbuj na CodePen**](https://codepen.io/gaearon/pen/ZXeOGM?editors=0010)

Reviewer: ```suggestion
[**Przetestuj kod na CodePen**](https://codepen.io/gaearon/pen/ZXeOGM?editors=0010)
```

> Podstawową regułą jest to że elementy wewnątrz wywołania `map()` potrzebują kluczy.

Reviewer 2: ```suggestion
Podstawową regułą jest to, że elementy wewnątrz wywołania `map()` potrzebują kluczy.
```

> ### Klucze muszą być unikalne tylko spośród rodzeństwa {#keys-must-only-be-unique-among-siblings}

Reviewer 2: ```suggestion
### Klucze muszą być unikalne tylko wśród rodzeństwa {#keys-must-only-be-unique-among-siblings}
```

Author: @jfrej Tutaj zostawiłbym tak jak było. Jak porównasz sobie:
https://sjp.pl/spo%C5%9Br%C3%B3d
z
https://sjp.pl/%C5%9Br%C3%B3d

To na pierwszy rzut oka, `wśród` gorzej leży na końcu zdania niż `spośród`. Dodatkowo `spośród` ma jedno znaczenie, które w tym przypadku bardzo dobrze pasuje.

@jakubdrozdek Jak sądzisz ?

Reviewer: Statystyki wystąpień _unikalny spośród_ vs. _unikalny wśród_ w Google "przeważają" w stronę _wśród_. Wydaje mi się, że różnica jest bardzo subtelna. Można by faktycznie doszukiwać się tutaj formy poprawnej, ale przeciętny czytelnik i tak nie zwróci uwagi i się nie przyczepi.

Moim zdaniem _Klucze muszą być unikalne tylko **spośród** rodzeństwa_ brzmi bardziej jakby "klucze" były jednym z elementów zbioru "rodzeństwo" (nasuwa mi się np. _Tylko 20% spośród wyborców zagłosowało na XYZ_ lub _Dwoje spośród trzech osób <coś tam zrobiło>_), co jest oczywiście nielogiczne w tym zdaniu. Natomiast _Klucze muszą być unikalne tylko **wśród** rodzeństwa_ wskazuje bardziej na to, na co ma wskazywać, czyli klucz jest pewną cechą każdego z "rodzeństwa" i musi być unikalny. Ale... może za bardzo mnie poniosło i różnica jest inna ;-) Przynajmniej tak mi się kojarzy, gdy to czytam na głos.

Author: Po dłuższym przemyśleniu **wśród** będzie lepsze. Naniosłem zmianę. :)

> Klucze używane wewnątrz tablic powinny być unikalne w kontekście swojego rodzeństwa. Jednakże nie muszą one być unikatowe globalnie. Możemy użyć tych samych kluczy gdy tworzymy dwie różne tablice:

Reviewer 2: ```suggestion
Klucze używane wewnątrz tablic powinny być unikalne w kontekście swojego rodzeństwa. Jednakże nie muszą one być niepowtarzalne globalnie. Możemy użyć tych samych kluczy, gdy tworzymy dwie różne tablice:
```

const posts = [
  {id: 1, title: 'Hello World', content: 'Welcome to learning React!'},

>   {id: 2, title: 'Installation', content: 'You can install React from npm.'}

Reviewer 2: Warto przetłumaczyć tytuły i treść.

Author: @jfrej Przetłumaczone proszę o review.

> ];id: 2, title: 'Installation', content: 'You can install React from npm.'}

Reviewer 2: Ta cała linia chyba została dodana przypadkowo
```suggestion
```

Author: Nie do końca cała a wszystko co jest po `];`
Naniosłem poprawkę. :)

];
ReactDOM.render(
  <Blog posts={posts} />,
  document.getElementById('root')
);
```
 

> [**Wypróbuj na CodePen**](https://codepen.io/gaearon/pen/NRZYGN?editors=0010)

Reviewer: ```suggestion
[**Przetestuj kod na CodePen**](https://codepen.io/gaearon/pen/NRZYGN?editors=0010)
```

> Klucze służą Reaktowi jako wskazówka, ale nie są one przekazywane do twoich komponentów. Jeżeli potrzebujesz tych wartości w swoim komponencie, przekaż je jawnie poprzez właściwość z inną nazwą:

Reviewer 2: Wygląda na to, że cała ta sekcja została skopiowana i wklejona poniżej. Jest przetłumaczona dwa razy.

Author: Poprawione. :)

 
```js{3,4}
const content = posts.map((post) =>
  <Post
    key={post.id}
    id={post.id}
    title={post.title} />
);
```
 
W powyższym przykładzie, komponent `Post` może odczytać `props.id`, ale nie `props.key`.
 
### Użycie map() wewnątrz JSX {#embedding-map-in-jsx}
 
W poprzednich przykładach zadeklarowaliśmy oddzielną zmienną `listItems` wewnątrz JSX:
 
```js{3-6}
function NumberList(props) {
  const numbers = props.numbers;
  const listItems = numbers.map((number) =>
    <ListItem key={number.toString()}
              value={number} />
  );
  return (
    <ul>
      {listItems}
    </ul>
  );
}
```
 
JSX pozwala na [wbudowanie każdego wyrażenia](/docs/introducing
ReactDOM.render(
  <Blog posts={posts} />,
  document.getElementById('root')
);
```
 
[**Wypróbuj na CodePen**](https://codepen.io/gaearon/pen/NRZYGN?editors=0010)
 

> Klucze służą Reaktowi jako wskazówka, ale nie są one przekazywane do twoich komponentów. Jeżeli potrzebujesz tych wartości w swoim komponencie, przekaż je jawnie poprzez właściwość z inną nazwą:

Reviewer: ```suggestion
Klucze służą Reaktowi jako wskazówka, ale nie są one przekazywane do twoich komponentów. Jeżeli potrzebujesz tych wartości w swoim komponencie, przekaż je jawnie poprzez właściwość o innej nazwie:
```

}
```
 

> JSX pozwala na [wbudowanie każdego wyrażenia](/docs/introducing-jsx.html#embedding-expressions-in-jsx) wewnątrz nawiasów klamrowych, dzięki temu możemy użyć wyniku funkcji `map()` bez przypisywania zmiennej:

Reviewer: ```suggestion
JSX pozwala na [wstawienie dowolnego wyrażenia](/docs/introducing-jsx.html#embedding-expressions-in-jsx) wewnątrz nawiasów klamrowych, dzięki temu możemy użyć wyniku funkcji `map()` bez przypisywania go do zmiennej:
```

}
```
 

> [**Wypróbuj na CodePen**](https://codepen.io/gaearon/pen/BLvYrB?editors=0010)

Reviewer: ```suggestion
[**Przetestuj kod na CodePen**](https://codepen.io/gaearon/pen/BLvYrB?editors=0010)
```

> Czasami dzięki temu kod jest bardziej przejrzysty, jednak ten sposób może być również nadużywany. Tak samo jak w JavaScript, wybór należy do Ciebie czy warto wyodrębnić zmienną aby zwiększyć czytelność. Zapamiętaj że jeżeli in ciało funkcji `map()` jest zbyt zagnieżdżone, to może być dobry czas aby [wyodrębnić komponent](/docs/components-and-props.html#extracting-components).

Reviewer 2: ```suggestion
Czasami dzięki temu kod jest bardziej przejrzysty, jednak ten sposób może być również nadużywany. Tak samo jak w JavaScripcie, wybór należy do Ciebie czy warto wyodrębnić zmienną aby zwiększyć czytelność. Zapamiętaj, że jeżeli ciało funkcji `map()` jest zbyt zagnieżdżone, to może być dobry czas aby [wyodrębnić komponent](/docs/components-and-props.html#extracting-components).
```

Author: Właśnie ciągle zastanawiałem się czy odmieniać `JavaScript`. Teraz już wiem. Dzięki!

> Czasami dzięki temu kod jest bardziej przejrzysty, jednak ten sposób może być również nadużywany. Tak samo jak w JavaScript, wybór należy do Ciebie czy warto wyodrębnić zmienną aby zwiększyć czytelność. Zapamiętaj że jeżeli in ciało funkcji `map()` jest zbyt zagnieżdżone, to może być dobry czas aby [wyodrębnić komponent](/docs/components-and-props.html#extracting-components).

Reviewer: ```suggestion
Czasami dzięki temu kod staje się bardziej przejrzysty, jednak łatwo doprowadzić do nadużywania tego sposobu. Tak samo jak w JavaScripcie, wybór należy do Ciebie, czy warto wyodrębnić zmienną, aby zwiększyć czytelność. Zapamiętaj, że jeżeli ciało funkcji `map()` jest zbyt zagnieżdżone, może to być odpowiedni moment na [wyodrębnienie komponentu](/docs/components-and-props.html#extracting-components).
```

Author: Połączyłem obie sugestie. 👍
